### ISSUE TITLE
Assemblers bind MultipleCorrectionGradient blind, so gradient-type patches lose quadratic exactness

### ISSUE THREAD

**Issue by ctdegroot** [DeGrootResearchGroup/aquaflux#467]
## What

`MultipleCorrectionGradient.bind` now takes an optional `boundary_gradient_weight`, and building the corrections through it is what keeps the scheme quadratic-exact when a field carries a gradient-type boundary condition (zero-gradient, Neumann, Robin). The two shipped assemblers do not pass it, so production still binds blind and is inexact on those patches.

## Why it matters

The reconstruction's first pass inverts `M1 - B` whenever a caller supplies the weight — which both assemblers do, on every reconstruction. The corrections are probed on `M1^-1`. So they correct an operator nobody evaluates, which is the mistake `_build_corrections` is written to avoid, and the scheme stops reproducing a quadratic.

Measured on an 8x8 perturbed quadrilateral grid (perturb 0.3, seed 3), `phi = 1.7 y^2 / 2`, x-normal walls zero-gradient — which that field satisfies exactly — and y-normal walls prescribed:

| binding | `OwnerGradient` | `SkewCorrectedGradient` |
|---|---|---|
| blind (what ships) | 2.009e-03 | 2.145e-03 |
| against the condition | 8.1e-15 | 2.6e-15 |

With every boundary value prescribed (`B = 0`) the blind binding is already exact at 2e-15, so the residue is entirely the boundary-condition operator's.

Pinned by `test_binding_against_a_boundary_condition_restores_quadratic_exactness`, which also pins that **both** halves of the repair are needed: probing the first pass through `(M1 - B)^-1` while still handing the closure the exact face value repairs `OwnerGradient` and makes `SkewCorrectedGradient` about five times *worse*.

## The work

- `ResidualAssembler` (`discretization/residual.py:296`) — one field, one set of conditions. It already computes the weight in `_boundary_gradient_weight`; the obstacle is that `bind` happens at build time and that function takes the state. For every shipped condition the gradient-dependence of a face value does not depend on the field, so evaluating it at a zero state is valid — but that is an assumption the code does not currently make anywhere, and it should be stated and tested rather than relied on silently.
- `MomentumContinuity` (`flow/momentum.py:299`) — the harder one. It binds **one** scheme and uses it for `u`, `v`, `w` and `p`, whose weights differ per field (for this duct, velocity's is nonzero on the outlet, pressure's on the inlet and walls). One binding cannot serve them, so this needs a binding per field.

## Not to do

Rebuilding the corrections inside `reconstruct` when a weight is supplied. It is correct but costs roughly seven extra face sums per reconstruction, per field, inside the Newton loop.

Two cheaper routes were tried and refuted, so they should not be re-derived:

- **Correcting only the defect**, which has a clean per-cell form `defect_B = (M1 - B)^-1 M1 defect`. It recovers about 40% of the gradient error (2.0e-3 -> 1.2e-3) and leaves the Hessian untouched at 1.6e-2, because `M2` is probed through the same wrong operator.
- **Iterating the reconstruction**, feeding the boundary values back at the corrected gradient. It diverges to a wrong fixed point (2.0e-3 -> 1.4e-1) because the weight already folds the gradient dependence in, so passing re-evaluated values alongside it double-counts that term.

**Comment by ctdegroot** [comment]
## Design findings, before implementation

Built on top of #468, which adds the `bind(mesh, geometry, boundary_gradient_weight)` capability this issue wires up. Branch `claude/issue-467-assembler-binding` exists off that.

### 1. The weight is exactly state-independent for every shipped condition

Measured on a 6x6 perturbed grid carrying `ZeroGradient`, `Neumann(flux=0.7)`, `Dirichlet(1.0)` and `Convective(h=2, t_inf=0.5)` simultaneously, with constant properties. `_boundary_gradient_weight` evaluated at six `(phi, gradient)` combinations — a zero state, a unit-normal random state, and one offset by 5.0, each at zero and at a random gradient:

```
max |w(state) - w(zeros)| over all six = 0.000e+00     (weight magnitude 2.6e-2)
```

So binding at build time with a zero state is **exact**, not an approximation, and this should be pinned by a test rather than left as a stated assumption. That removes the obstacle this issue originally recorded for `ResidualAssembler`.

### 2. But it is not state-independent of the *properties*, and that is the real wrinkle

`boundary_values` reads `properties.get(self.coefficient, zeros)` — with a **zeros default**. So the weight can be computed with an empty property map without raising, and for `Dirichlet`, `ZeroGradient` and `Neumann` the gradient-dependent part is the geometric tangential offset, which does not involve `Gamma`. For `Convective` the face value blends `h` and `Gamma`, so its gradient coefficient does involve `Gamma`, and computing the weight against the zeros default would give a **silently wrong** weight — corrections built for an operator that is not the one applied, which is the exact failure mode this whole line of work is about.

`PropertyModel.evaluate(cell_zones, fields)` is fine with `fields={}` for constant and per-zone properties, and cannot be evaluated at build time for a **calculated** property.

So the implementation has to decide, explicitly, what to do when the properties cannot be evaluated at build time. Binding blind and staying silent is what ships today and is the thing being fixed; binding blind with a warning is the safe default; raising is defensible since the caller can pass the weight itself. **This is the decision to make first — it should not be made implicitly by whichever branch happens to be taken.**

### 3. Shape of the change

- `ResidualAssembler.build` — two-phase: construct with the blind binding, compute `weight` from the constructed assembler (it needs `self.boundary` and `self.mesh`), then `dataclasses.replace(assembler, gradient_scheme=scheme.bind(mesh, geometry, weight))`. Small, contained.
- `MomentumContinuity` — the harder half, unchanged from the original description: it binds **one** scheme and uses it for `u`, `v`, `w` and `p`, whose weights differ per field (on the tetrahedral duct, velocity's is nonzero on the outlet and pressure's on the inlet and walls). It needs a binding per field, which changes the shape of the class. Note other sessions' probes subclass and read `momentum.gradient_scheme`, so this half wants coordinating.

### 4. Not a confound for #435, established separately

The boundary probe/apply gap cannot explain the tetrahedral march failure, by two independent arguments now: the second pass withheld on rings <= 2 leaves active only cells whose own faces and whose neighbours' first passes are all interior, so `B` is zero throughout their construction, and that arm still fails; and the outlet-cell arms (first pass on outlet cells only; on inlet+outlet; rings <= 2 plus inlet/outlet; rings <= 3 with outlet kept full) came back refuting the outlet confound directly. This issue is worth doing on its own merits, not as a #435 fix.

**Comment by ctdegroot** [comment]
## `ResidualAssembler` half is done; here is the spec for the `MomentumContinuity` half

Done on branch `claude/issue-467-assembler-binding` (commit `517ed2f`, on top of #468): `build` binds against the conditions, the scalar path is exact end to end (2.009e-3 -> 8.1e-15), a property needing a field raises and names `boundary_gradient_weight=`, and `GradientScheme.bind` carries the weight as an optional third argument so every scheme binds through one signature.

### What the flow half needs, with the facts checked

Both weights already exist and are the right shape — this is a wiring and lifetime problem, not a derivation one:

- `MomentumContinuity._velocity_boundary_gradient_weight(velocity, component, grad_velocity)` -> `(n_faces, dim)`, one `jax.jvp` per gradient component.
- `MomentumContinuity._pressure_boundary_gradient_weight(pressure, grad_pressure)` -> `(n_faces, dim)`, likewise.

They differ per field, which is the whole obstacle: on the tetrahedral duct the velocity weight is nonzero on the outlet and the pressure weight on the inlet and walls. One bound scheme cannot serve both, so `build` needs **`dim + 1` bindings** — one per velocity component, one for pressure — each built from its own weight evaluated at a zero state.

Evaluating at a zero state is exact, not approximate, for the same reason it is in `ResidualAssembler`, and that is now pinned by `test_the_boundary_gradient_weight_does_not_depend_on_the_state`. Worth re-pinning for the momentum weights specifically, since they are computed by different code.

### The part that needs a decision, not just typing

`MomentumContinuity.gradient_scheme` is read from outside the class, so replacing it changes a surface other code depends on:

- `flow/initialization.py:258` (`potential_flow`)
- `turbulence/initialization.py:109`, via the turbulence model's own scheme
- `parallel/distributed.py:268` reads `assembler.gradient_scheme` to decide whether a halo exchange is needed
- other sessions' probes construct `MomentumContinuity.build(..., gradient_scheme=...)` and some wrap what comes back

So the choice is between keeping `gradient_scheme` as the blind binding and adding the per-field ones beside it (compatible, but two spellings of one concept — the kind of duplicate surface this project treats as a defect), or replacing it with the per-field collection and updating every reader (cleaner, wider blast radius). **I would take the second**, but it should be a deliberate call rather than whichever is convenient, and it wants coordinating with the #435 investigation, whose probes subclass and read this class while it is running.

### Not a confound for #435

Recorded again because it keeps coming up: the boundary probe/apply gap cannot explain the tetrahedral march failure. Two independent arguments now — the rings <= 2 arm leaves active only cells whose own faces and whose neighbours' first passes are all interior, so `B` is zero throughout their construction and that arm still fails; and the outlet-cell arms refuted the outlet confound directly. Do this for correctness, not as a fix for that.
